### ISSUE TITLE
Adjusting a mixture of Gaussians to an image (loss goes nan)

### ISSUE THREAD

**Issue by vianamp** [keras-team/keras#8791]
Hi guys.

I am trying to find the Gaussian Mixture Model (GMM) that best fits the gray levels found in a given input image. The network outputs the GMM parameters and my loss function is the MSE between i) the image created by the GMM with the predicted parameters and ii) the original image. The issue is that the loss is becoming nan after the first iteration. Any ideas why? Also, in general the loss function returns a single value or a tensor with a value per sample in the batch?

Many thanks,

```python
import numpy as np
import tensorflow as tf
from keras import backend as K
from keras.models import Model
from keras.layers import Input, Conv2D, Flatten, Dense, concatenate

IMAGE_D = 1
IMAGE_W = 28
IMAGE_H = 28
BATCH_SIZE = 8
NGAUSSIANS = 5

def kernel_2d(x1,x2,mu1,mu2,s1,s2,rho):
	
	print('x1:',x1.shape)
	print('mu1:',mu1.shape)
	
	norm1 = tf.subtract(x1,mu1)
	norm2 = tf.subtract(x2,mu2)
	print('norm1:',norm1.shape)
	
	s1s2 = tf.multiply(s1,s2)
	print('s1s2:',s1s2.shape)
	
	sq1 = tf.square(tf.divide(norm1,s1))
	sq2 = tf.square(tf.divide(norm2,s2))
	print('sq1:',norm1.shape)
	
	norm12 = tf.multiply(norm1,norm2)
	print('norm12:',norm12.shape)
	
	sq12 = 2 * tf.multiply(rho,tf.divide(norm12,s1s2))
	print('sq12:',sq12.shape)
	
	z = sq1 + sq2 - sq12
	print('z:',z.shape)
	
	negRho = 1 - tf.square(rho)
	print('negRho:',negRho.shape)
	
	result = tf.exp(tf.divide(-z, (2*negRho)))
	print('result:',result.shape)
	
	denom = 2*np.pi*tf.multiply(s1s2, tf.sqrt(negRho))
	print('denom:',denom.shape)
	
	result = tf.divide(result, denom)
	print('result:',result.shape)
	
	return result

def mse_gmm(img_true):

	def get_img_from_gmm(y_true, y_pred):

		lmi = tf.expand_dims(y_pred[:,(0*NGAUSSIANS):(1*NGAUSSIANS)], axis=1)
		mu1 = tf.expand_dims(y_pred[:,(1*NGAUSSIANS):(2*NGAUSSIANS)], axis=1) * IMAGE_W
		mu2 = tf.expand_dims(y_pred[:,(2*NGAUSSIANS):(3*NGAUSSIANS)], axis=1) * IMAGE_H
		sg1 = tf.expand_dims(y_pred[:,(3*NGAUSSIANS):(4*NGAUSSIANS)], axis=1)
		sg2 = tf.expand_dims(y_pred[:,(4*NGAUSSIANS):(5*NGAUSSIANS)], axis=1)
		rho = tf.expand_dims(y_pred[:,(5*NGAUSSIANS):(6*NGAUSSIANS)], axis=1)

		coords = np.array(np.meshgrid(np.arange(IMAGE_W),np.arange(IMAGE_H)), dtype=np.float32).T.reshape(-1,2,1)
		coords = coords.repeat(NGAUSSIANS, axis=-1)
		coords = np.expand_dims(coords, axis=0).repeat(BATCH_SIZE, axis=0)
		print('coords:',coords.shape)

		x1 = K.variable(coords[:,:,0])
		x2 = K.variable(coords[:,:,1])

		img_pred = kernel_2d(x1=x1, x2=x2, mu1=mu1, mu2=mu2, s1=sg1, s2=sg2, rho=rho)
		img_pred = tf.multiply(img_pred,lmi)
		img_pred = tf.reduce_sum(img_pred, axis=-1)
		img_pred = tf.reshape(img_pred,(-1,IMAGE_W,IMAGE_H,1))

		print('img_pred:',img_pred.shape)
		print('img_true:',img_true.shape)

		err = tf.reduce_mean(tf.reduce_mean(tf.square(img_pred-img_true), axis=2), axis=1)
		print('err:',err.shape)
		
		return err

	return get_img_from_gmm

#
# MODEL
#

input_image = Input(shape=(IMAGE_W,IMAGE_H,IMAGE_D))
conv1_out = Conv2D(filters=128, kernel_size=5, strides=(2,2))(input_image)
conv2_out = Conv2D(filters=256, kernel_size=4, strides=(2,2))(conv1_out)
conv3_out = Conv2D(filters=512, kernel_size=3, strides=(2,2))(conv2_out)
conv4_out = Conv2D(filters=1024, kernel_size=2, strides=(2,2))(conv3_out)
flat_output = Flatten()(conv4_out)
dense1_out = Dense(units=1024, activation='relu')(flat_output)
dense2_out = Dense(units=512, activation='relu')(dense1_out)
out_lambda = Dense(units=1*NGAUSSIANS, activation='softmax')(dense2_out)
out_mu = Dense(units=2*NGAUSSIANS, activation='relu')(dense2_out)
out_sigma = Dense(units=2*NGAUSSIANS, activation='tanh')(dense2_out)
out_rho = Dense(units=1*NGAUSSIANS, activation='relu')(dense2_out)
output = concatenate([out_lambda, out_mu, out_sigma, out_rho])
model = Model(inputs=[input_image], outputs=output)
model.compile(optimizer='sgd', loss=[mse_gmm(input_image)])

print(model.summary())

#
# TRAIN
#

X = np.random.rand(BATCH_SIZE,IMAGE_W,IMAGE_H,IMAGE_D)

DUMMY_Y = np.zeros((BATCH_SIZE,6*NGAUSSIANS))

model.fit(x=[X], y=[DUMMY_Y], epochs=4)
```

```
Epoch 1/4
2017-12-14 17:49:27.968734: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
8/8 [==============================] - 1s 77ms/step - loss: 3.9720
Epoch 2/4
8/8 [==============================] - 0s 11ms/step - loss: nan
Epoch 3/4
8/8 [==============================] - 0s 11ms/step - loss: nan
Epoch 4/4
8/8 [==============================] - 0s 10ms/step - loss: nan
```

**Comment by gabrieldemarmiesse** [comment]
Hi!
I don't really know why NaN appears. My guess is that the learning rate is too high. It happens often. 
But for your question about the loss, keras suppose that the loss function is going to return a tensor of rank 0 per batch (so a single scalar per batch).
Btw, the keras guidelines regarding github issues advise you to post this question on stackoverflow rather than here. People here are looking for bugs in the keras codebase. You'll have more answers in stackoverflow! 
I hope it helps.
